[RFC] arm32: introduce suspend/resume common functions

MrVan opened the pull request:
This is to introduce low power suspend/resume support for arm32. The work flow is similar as linux suspend/resume, the different is linux suspend/resume saves/restores linux side contexts and system registers, while the optee suspend/resume saves/restores tee side contexts and system registers. Also some differences that ttbr0 is saved/restored. 
Now only 1 cluster max 4 cores supported and with Cortex-A7/A9 MP core tested. 
This piece code is not the whole picture, but these are all the common code I implemented, others are platform specific code, such as platform normally works at svc mode when reset, but when tee resume we need to switch to monitor mode because psci frameworks runs in monitor mode. And code handling platform suspend/resume coping into onchip ram that could survive cpu suspend/resume.

MrVan commented:
Are there any comments about this RFC patch set?

igoropaniuk commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,261 @@
+	read_nmrr r5
+	read_vbar r6
+	read_nsacr r7
+	stmia	r0, {r4 - r7}
```
stmia r0**!**, {r4 - r7} ?

MrVan replied:
There is no need to use r0! here, this is the last inst to store register to stack and then r0 is not needed.

igoropaniuk commented on core/arch/arm/include/sm/pm.h:
```diff
@@ -0,0 +1,40 @@
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
```
header guards

MrVan replied:
Fix in V2

jforissier commented:
@MrVan FYI the Travis build failure can be fixed by rebasing on top of the current master branch, in order to to get 9aea3dc83a96 ("Travis: add -p to mkdir $HOME/bin").

MrVan commented:
Patch updated with addressing @igoropaniuk 's comments and rebased with 9aea3dc ("Travis: add -p to mkdir $HOME/bin").  Any comments?

etienne-lms commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,261 @@
+	 * the &sleep_save_sp_virt = ((u32)suspend_sp + 4 * cpu_id)
+	 */
+	and	r6, r6, #0x3
+	lsl	r6, r6, #0x3
```
`lsl	r6, r6, #0x2` ?

MrVan replied:
No. Each cpu needs 2 words(8 bytes), `lsl r6, r6, #3`, means r6 * 8

etienne-lms commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,261 @@
+	ldr	r3, =cpu_resume_after_mmu
+	isb
+	write_sctlr r0
+	mrc     p15, 0, r0, c0, c0, 0
```
Could you cross check this (from here to the 2nd `mov r0, r0` below. Looks useless.

MrVan replied:
`write_sctlr r0` actually will enable MMU. Yeah, that could removed on ARMV7 cpus. This could be
```
write_tlbiall
write_sctlr r0
dsb
isb
``

etienne-lms commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,261 @@
+	 */
+	read_diag r10
+	teq     r5, r10
+	mcrne   p15, 0, r5, c15, c0, 1
```
Here and below: is it harmful to restore an already updated value ?

MrVan replied:
No. This could be simplified. No one actually restore it before this point.

MrVan commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,261 @@
+.section .text
+FUNC cpu_suspend, :
+UNWIND( .fnstart)
+	stmfd	sp!, {r4 - r11, lr}
```
I'll try this. Thanks.

MrVan commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,261 @@
+	ldr	r4, =suspend_reg_size
+	mov	r5, sp
+	add	r4, r4, #12
+	sub	sp, sp, r4
```
This is to reserve space in stack for saving regs that will lost when cpu losing power. unwind may not able to handle this. I may need to add KEEP_PAGER the functions.

MrVan commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,261 @@
+	 */
+	and	r6, r6, #0x3
+	lsl	r6, r6, #0x3
+	stmfd	sp!, {r0, r1}
```
ok

MrVan commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,261 @@
+	ldr	r4, =suspend_reg_size
+	add	r4, r4, #12
+	add	sp, sp, r4
+	ldmia	sp!, {r4 - r11, lr}
```
ok

jenswi-linaro commented on core/arch/arm/sm/pm.c:
```diff
@@ -0,0 +1,89 @@
+ * Each cpu needs 2 words to save ptr and virt_to_phys(ptr).
+ * We max support 4 cpus now,so need 4 * 2 * sizeof(uint32_t);
+ */
+uint32_t suspend_sp[4 * 2] __aligned(8);
```
This would be more convenient to store in `struct thread_core_local`. What's the alignment for?

MrVan replied:
I am not sure. it maybe a bit hard to refer thread_core_local, i'll try. No specific requirement on the alignment, could be removed.

MrVan replied:
After a thought, thread_core_local could not be used here. Because suspend_sp will be used when cpu resume at very early stage to restore registers from sp(physical address). At this time,accessing thread_core_local is not easy,  also have to fix the layout of thread_core_local for the saved sp pointer. I prefer use suspend_sp here.

jenswi-linaro replied:
`thread_core_local` could be made a global variable just as `suspend_sp` is here, I don't see what's different.
`core/arch/arm/kernel/asm-defines.c` can be used to find offsets of different elements in the struct.

MrVan replied:
ok. I'll try this. please hold on for the review patch, I missed this comments when updating the patch.

jenswi-linaro commented on core/arch/arm/sm/pm.c:
```diff
@@ -0,0 +1,89 @@
+	 * The content will be passed to cpu_do_resume
+	 * as register r1 and sp
+	 */
+	*ptr++ = 0; /* Not used now */
```
Use a proper struct for this instead.

MrVan replied:
I would not add a struct here. The layout will be refered in asm code, It is clear using current method.

jenswi-linaro replied:
It's not entirely clear to me.
We use structs at several places in the code that are accessed from assembly code, both for convenience and as documentation. There's `core/arch/arm/kernel/asm-defines.c` to help if needed.

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,261 @@
+ */
+.align 2
+.global	suspend_reg_size
+.equ	suspend_reg_size, 4 * 16
```
Couldn't a regular define be used instead?

MrVan replied:
I'll give a try.

MrVan commented:
@jenswi-linaro @etienne-lms Update with your review comments addressed. Please check whether this is ok for you. I have not do board test for the new review patch, later I'll test it.

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,244 @@
+	read_mpidr r1
+	and	r1, r1, #0x3
+
+	adr	r0, _suspend_sp
```
`ldr r0, =suspend_sp` instead?

If you we're using `thread_core_local` instead the sequence would be something like (and then we'd get rid of the assumption above too):
```
bl get_core_pos
mul r0, r0, #THREAD_CORE_LOCAL_SIZE
ldr r1, =thread_core_local
add r1, r1, r0
```

MrVan replied:
At this point mmu is not enabled. bl get_core_pos could not be used here. I'll use thread_core_local here.

jenswi-linaro replied:
Isn't `get_core_pos()` flat mapped?

MrVan replied:
I could not rely on physical address equals to virtual address. I could use bl get_core_pos here, but when virtual address is not same physical address, this needs to rewrite. I could add a piece code just as get thread_core_local phyiscal address to get get_core_pos physical address here.

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,244 @@
+#include <keep.h>
+#include <kernel/unwind.h>
+
+#define PRRR			0x034
```
Unused defines?

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,244 @@
+FUNC cpu_resume, :
+UNWIND(.fnstart)
+	/* Disable IRQ/FIQ */
+	mrs	r1, cpsr
```
`cpsid if` is only one instruction. I'd recommend disabling async aborts too, so then it would be `cpsid aif`.

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,244 @@
+	write_sctlr r0
+	dsb
+	isb
+	nop
```
Any reference to why you need exactly three nops?

MrVan replied:
No specific reference, just would like to flush instruction pipe.

jenswi-linaro replied:
Why is that needed? Handling an erratum?

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,244 @@
+	adr	lr, cpu_suspend_panic
+	/* Jump to arch specific suspend*/
+	pop	{r0, pc}
+	/* cpu not suspended */
```
Can this happen, considering that `lr` is pointing to `cpu_suspend_panic()`

MrVan replied:
This is bug here. This will be fixed in new version patches.

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,244 @@
+	blx	cpu_suspend_save
+	adr	lr, cpu_suspend_panic
+	/* Jump to arch specific suspend*/
+	pop	{r0, pc}
```
Where is the value of `pc` supposed to be written?

MrVan replied:
cpu_suspend(arg, platform_suspend_func), Here pc is platform_suspend_func.

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,244 @@
+	push	{r4, r5}
+	read_midr r4
+	ubfx	r5, r4, #4, #12
+	ldr	r4, =0xc07
```
This magic number could be a defined in `<arm32.h>` instead.
The assignment could also use `mov_imm` instead.

MrVan replied:
Using mov_imm, "Error: invalid constant (c07) after fixup"

MrVan commented:
@jenswi-linaro I updated the patches, not review patch. I have tested it on mx7dsabresd, with suspend and dual-core low power idle, works well.  I have switched to use thread_core_local and addresses a few bugs in the previous review patch which not tested.

jenswi-linaro commented on core/arch/arm/include/sm/pm.h:
```diff
@@ -0,0 +1,49 @@
+};
+
+/* suspend/resume core functions */
+void cpu_suspend_init(void);
```
I'd prefer a "sm_pm" prefix over a "cpu" prefix for these functions.

MrVan replied:
This function is just a place holder, I could remove it, since not used now. Actually I would prefer cpu means this piece code is only related to cpu core, not about device in a SoC. This function will be removed in new version patches.

jenswi-linaro commented on core/arch/arm/sm/pm.c:
```diff
@@ -0,0 +1,77 @@
+#include <sm/sm.h>
+#include <sm/pm.h>
+
+#include "../kernel/thread_private.h"
```
Please export `struct thread_core_local` in `<thread.h>` instead of including `"../kernel/thread_private.h"` here.

MrVan replied:
ok

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,224 @@
+
+FUNC cpu_v7_suspend, :
+UNWIND(.fnstart)
+	push	{r4 - r12, lr}
```
No need to save `r12` and `lr`

MrVan replied:
r12 may not needed, but lr is needed. cpu_suspend_save->cpu_do_suspend->cpu_v7_suspend. lr is needed to back to cpu_suspend_save. r12 here is just make stack 8 bytes aligned.

jenswi-linaro replied:
I see no calls in this function

MrVan replied:
ok. I'll refine this.

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,224 @@
+#include <kernel/unwind.h>
+
+.section .text
+FUNC cpu_suspend, :
```
Please add a comment describing arguments and return value of each exported function in this file.
Something like:
```
/*
 * int cpu_suspend(uint32_t arg, int (*fn)(uint32_t));
 */
```
Is enough. It doesn't hurt to describe the local functions in the same manner but it's not as important.

MrVan replied:
ok

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,224 @@
+.section .text
+FUNC cpu_suspend, :
+UNWIND( .fnstart)
+	push	{r4 - r12, lr}
```
It looks like more registers than needed are pushed.

MrVan replied:
Yeah. It makes me easy to use registers. Let's just keep it?

jenswi-linaro replied:
It makes it harder to read the code since you start to wonder to what purpose those other registers are pushed. Only push what's needed, that's easiest for the ones reading the code.

MrVan replied:
Rethought on this. need to save `r4-r12,lr`, when back to caller of cpu_suspend, need to keep register not changed. During cpu_suspend, cpu may power off and context lost. So I need to save them and restore them in cpu_resume.

jenswi-linaro replied:
`cpu_suspend_save()` is supposed to restore whatever registers when returning and no other functions are called here. I don't see why all registers need to be saved.

MrVan replied:
See the example:
```
int psci_suspend_function()
{
        fun1();
        rc = cpu_suspend(a, platform_specific_suspend);
        func2();
}
```
cpu_suspend -> platform_specific_suspend -> cpu suspended, no power, all registers lost.

Now we trigger cpu resume by an interrupt, so the flow is:
platform_specific_resume->cpu_resume->back to func2, the following instruction after cpu_suspend call.

The important point is cpu will lose power, I have to push the registers on stack(dram), and when resume, restore the registers from stack, not because no other functions called here.

jenswi-linaro replied:
I'm sorry, I can't review this. I'm missing too much of the whole picture to make a proper review.

MrVan replied:
This is the linux side code, the secure side code has similar strategy.
(1) http://elixir.free-electrons.com/linux/v4.13-rc6/source/arch/arm/kernel/sleep.S#L60
save registers when suspending.
(2) http://elixir.free-electrons.com/linux/v4.13-rc6/source/arch/arm/kernel/sleep.S#L94
Restore registers when suspend aborted.
(3) http://elixir.free-electrons.com/linux/v4.13-rc6/source/arch/arm/kernel/sleep.S#L116
Restore registers when cpu resume from power off state
(4) http://elixir.free-electrons.com/linux/v4.13-rc6/source/arch/arm/mach-imx/suspend-imx6.S#L348  
platform resume point invoke cpu_resume and jump to (3).

In Secure side, I have similar strategy to implement suspend/resume feature. The platform specific side code has not been cleaned up. After the common code is ok, I'll send i.mx specific asm code out. From the linux code, you could understand how it work.s

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,224 @@
+UNWIND( .fnend)
+END_FUNC cpu_suspend
+
+FUNC cpu_v7_suspend, :
```
This seems to be a local function, use the `LOCAL_FUNC` macro instead.

MrVan replied:
ok

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,224 @@
+	cmp	r5, r4
+	beq	a7_suspend
+	/*
+	 * TODO: add more Cortex A[x] check?
```
Yes, reporting an error instead of treating everything else as Cortex A9 would be helpful.

MrVan replied:
This code supports A7 and A9. I do not have other platforms to test.  I'll check to add code to report error for unsupported platforms.

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,224 @@
+
+	adr	r0, _suspend_sp
+	ldr	r2, [r0]
+	/* Get the physical address of thread_core_local */
```
I don't understand this comment, how are these three instructions different from the more direct approach:
```
ldr r0, =thread_core_local
ldr r0, [r0]
```
It seems I'm missing a part of the big picture.

MrVan replied:
When cpu suspended, it will lose power. So in the resume path, before mmu is enabled, we have to access physical address. ldr r0, =thread_core_local, gives virtual address to r0, I understand that virtual address equals to physical address to tee core now. But it may change in future, the piece code could handle that.

Let me explain more, the address of _suspend_sp contains the value `thread_core_local virtual address - _suspend_sp virtual address`.
```
.align 2                                                                        
.extern thread_core_local                                                       
_suspend_sp:                                                                    
        .long   thread_core_local - .
```

So when resume, mmu has not been enabled, I could use the offset + _suspend_sp phyiscal address to get the thread_core_local physical address.

THen why `adr r0, _suspend_sp`? adr is used for relative addressing mode, if I use ldr r0,=xxx, then absolute addressing mode is used.

In the resume path, I need to use relative addressing mode, so use `adr r0, _suspend_sp` to get the physical address of _suspend_sp, then _suspend_sp + (offset with thread_core_local) to get physical address of thread_core_local.

jenswi-linaro replied:
Let's skip preparing for non-linear mapping here to focus on one issue at a time.

MrVan replied:
I am thinking to use non-linear mapping and preparing one image support i.MX family. This piece code is not that easy to debug, if using current method, I would not spend efforts on debuging suspend/resume issues after migrating to use non-linear mapping. Hope you could agree on this. If code is not clear, I could add more comments in code.

jenswi-linaro replied:
OK, additional comments will work for me when reviewing.

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,224 @@
+
+	read_actlr r4
+	teq	r4, r9
+	mcrne   p15, 0, r9, c1, c0, 1
```
What difference does it make to write the same value again?

MrVan replied:
I saw linux kernel code has similar strategy. I do not have specific reason on this.

jenswi-linaro replied:
Since we don't understand why we should do it, let's skip it.
Doing stuff without knowing why is like cargo cult programming, https://en.wikipedia.org/wiki/Cargo_cult_programming

MrVan replied:
ok.

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,224 @@
+FUNC cpu_resume_mmu, :
+UNWIND(.fnstart)
+	ldr	r3, =cpu_resume_after_mmu
+	write_tlbiall
```
TLB was already invalidated in `cpu_do_resume()`, why is that needed again?

MrVan replied:
Not needed. I'll drop it.

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,224 @@
+	nop
+	nop
+	nop
+	bx	r3
```
Why not just a simple `b cpu_resume_after_mmu`?

MrVan replied:
write_sctlr r0 will enable MMU. After mmu enabled, bx r3 to jump to cpu_resume_after_mmu.

jenswi-linaro replied:
The address you jump to will be the same so why make it complicated?

MrVan replied:
ok. I'll try it.

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,224 @@
+UNWIND( .fnend)
+END_FUNC cpu_resume
+
+FUNC cpu_resume_after_mmu, :
```
`LOCAL_FUNC`

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,224 @@
+UNWIND( .fnend)
+END_FUNC cpu_resume_after_mmu
+
+FUNC cpu_resume_mmu, :
```
`LOCAL_FUNC`

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,224 @@
+UNWIND( .fnend)
+END_FUNC cpu_do_suspend
+
+FUNC cpu_resume, :
```
`LOCAL_FUNC`

MrVan commented on core/arch/arm/include/arm32.h:
```diff
@@ -33,6 +33,8 @@
 #include <stdint.h>
 #include <util.h>
 
+#define A7_PART_NUM	0xC07
```
ok

MrVan commented:
@jenswi-linaro Updated with a single review patch. I understand code is a bit hard to read, thanks for your detailed reviewing. Patch has tested on mx7dsabresd board, with suspend/resume low power idle and xtest.

MrVan commented:
Any more comments on this? In my side, it passed 24 hours test on mx7dsabresd.

jenswi-linaro commented:
I guess we're all waiting for the remaining patches to get the full picture.

MrVan commented:
ok. I'll clean up platform specific code and send out as whole for review.

MrVan commented:
Updated the patchset to include platform specific code.  The PATCH 9/9 is not good enough to upstream, if ok, I would prefer other 8 patches go upstream and leave the last patch. welcome comments.

jbech-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,229 @@
+	ldr	r2, [r0]
+	add	r0, r0, r2
+
+	/* Assmue max 4 cpus and 1 cluster */
```
Since this is generic code, can't we try to support more cpus and more than 1 cluster from start? 4x2 seems pretty common these days.

MrVan replied:
@jbech-linaro This patchset is to support armv7 which mostly only contains 4 cores. Regarding armv8 aarch32, I do not have platforms to test. I could use simliar method as accessing thread_core_local here to use get_core_pos.

jbech-linaro replied:
True, well I think this would serve as a good start and we can deal with that question later on. I'm not familiar with imx-specfics, but all in all I think this looks good, thanks!

jenswi-linaro commented on core/arch/arm/plat-imx/pm/imx7_suspend.c:
```diff
@@ -0,0 +1,183 @@
+
+	/*
+	 * Store sec ctx regs
+	 * Now we are in SVC sec mode
```
Is this comment accurate? I can't find where the switch to SVC sec mode is done.

MrVan replied:
This should be monitor mode. Also in the asm mode switch code, it should be back to monitor code and save svc mode, as the following:
```
-               "mov    r3, #0x16\r\n" /* MON */
+               "mov    r3, #0x13\r\n" /* SVC */
                "orr    r3, r3, r4\r\n"
                "msr    cpsr, r3\r\n"
                "mrs    r1, spsr\r\n"
                "str    r1, [r2], #4\r\n"
                "str    sp, [r2], #4\r\n"
                "str    lr, [r2], #4\r\n"
-               "msr    cpsr, r0\r\n"           /* SVC mode */
+               "msr    cpsr, r0\r\n"   /* Mon mode */
```
When pushed to github, missed to fix this.

jenswi-linaro commented on core/arch/arm/plat-imx/pm/imx7_suspend.c:
```diff
@@ -0,0 +1,183 @@
+	 * Store sec ctx regs
+	 * Now we are in SVC sec mode
+	 */
+	asm volatile (
```
This looks a lot like `sm_save_modes_regs() is doing`, couldn't `sm_save_modes_regs(&nsec->mode_regs)` be used instead?

MrVan replied:
I'll check that. After reviewing patch 9/9, you could get the whole picture of how it works:)

jenswi-linaro commented on core/arch/arm/plat-imx/pm/imx7_suspend.c:
```diff
@@ -0,0 +1,183 @@
+	regs = &reg_stack[0];
+
+	/*
+	 * Store sec ctx regs
```
It's actually nonsecure ctx regs.

MrVan replied:
non secure regs are saved by non-secure world by itself. This is to save secure ctx, because cpu lose power, all regs will be lost, so we need to save secure ctx.

jenswi-linaro replied:
The secure registers hasn't been loaded by the secure monitor yet so the content is actually non-secure, unless I'm missing something.

MrVan replied:
I misunderstand. yes. secure registers not loaded. This is save non-secure context. I'll correct the comment.

jenswi-linaro commented on core/arch/arm/plat-imx/pm/imx7_suspend.c:
```diff
@@ -0,0 +1,183 @@
+	plat_cpu_reset_late();
+
+	/* Restore register of different mode in secure world */
+	asm volatile (
```
If `sm_save_modes_regs()` is used above, use `sm_restore_modes_regs(&nsec->mode_regs)` instead.

MrVan replied:
This is restore secure world context. Ditto, I'll check using sm_restore_modes_regs.

jenswi-linaro commented on core/arch/arm/plat-imx/pm/psci-suspend-imx7.S:
```diff
@@ -0,0 +1,733 @@
+	str	r7, [r11, #0x38]
+wait_shutdown:
+	wfi
+	nop
```
Are the `nop` instructions really needed?

MrVan replied:
This similar code has been in FSL/NXP releases for long time. I do not want to remove them for potential risk here.

jenswi-linaro commented on core/arch/arm/plat-imx/pm/psci-suspend-imx7.S:
```diff
@@ -0,0 +1,733 @@
+wfi:
+	/* Enter stop mode */
+	wfi
+	nop
```
Are all these `nop` instructions needed?

MrVan replied:
Ditto.

jenswi-linaro commented on core/arch/arm/plat-imx/pm/psci-suspend-imx7.S:
```diff
@@ -0,0 +1,733 @@
+	mov	r4, #(BIT32(6) | BIT32(7))
+	orr	r3, r3, r4
+	msr	cpsr, r3
+	nop
```
`nop` for a particular reason?

MrVan replied:
No. I'll remove it.

jenswi-linaro commented on core/arch/arm/plat-imx/pm/psci-suspend-imx7.S:
```diff
@@ -0,0 +1,733 @@
+	bne     _inv_nextway
+
+	dsb                             @ ; synchronise
+	nop
```
Wouldn't an `isb` instead of these two `nop` be more appropriate?

MrVan replied:
yes.

jenswi-linaro commented on core/arch/arm/plat-imx/pm/psci-suspend-imx7.S:
```diff
@@ -0,0 +1,733 @@
+
+resume:
+	write_iciallu
+	write_bpiall
```
Should perhaps have `dsb` `isb` after this?

MrVan replied:
ok. will add them.

jenswi-linaro commented on core/arch/arm/plat-imx/pm/psci-suspend-imx7.S:
```diff
@@ -0,0 +1,733 @@
+	isb
+
+	/* switch to monitor mode from svc mode */
+	mov	r3, #CPSR_MODE_MON
```
`cps #CPSR_MODE_MON` instead of this long sequence?

MrVan replied:
ok

MrVan commented:
@jbech-linaro Updated with using get_core_pos in suspend.S @jenswi-linaro Updated with addressing comments for psci-suspend-imx7.S. 

Sending out the imx specific things to let people could get a whole picture, but here the first thing I would move forward is upstreaming the common parts, saying patch 1/9-8/9.  I am glad, if you think the last patch 9/9 is good to be accepted. In next version patch, I will extract the psci macros into standalone patch from patch 9/9.

jenswi-linaro commented on core/arch/arm/plat-imx/pm/imx7_suspend.c:
```diff
@@ -54,64 +53,15 @@ int imx7_cpu_suspend(uint32_t power_state __unused, uintptr_t entry,
-		: "r0", "r1", "r2", "r3", "r4"
-		);
+	/* Store non-sec ctx regs */
+	sm_save_modes_regs(&mode_regs[0]);
```
`&nsec->mode_regs` is not used at this stage and could be used instead of the local variable `mode_regs`.

MrVan replied:
Thanks. I'll use that in next version patch after receiving more comments.

etienne-lms commented on core/arch/arm/include/mm/core_mmu.h:
```diff
@@ -481,6 +481,8 @@ static inline TEE_Result cache_op_outer(enum cache_op op __unused,
 /* Check cpu mmu enabled or not */
 bool cpu_mmu_enabled(void);
 
+void map_memarea_sections(const struct tee_mmap_region *mm, uint32_t *ttb);
```
This is core_mmu_v7.c, specific not LPAE. A generic `cpu_mmu_map_memarea()` could direct to the mmu_v7/lpae map functions. Lpae would be not supported at first I guess.
minor: can add a description since the function is now exported.

MrVan replied:
I do not have platforms with LPAE enabled. I'll add a description.

etienne-lms commented on core/arch/arm/plat-imx/conf.mk:
```diff
@@ -78,4 +78,8 @@ $(call force,CFG_SECURE_TIME_SOURCE_REE,y)
 
+ifeq ($(filter y, $(CFG_PSCI_ARM32)), y)
+CFG_CORE_RWDATA_NOEXEC ?= n
+endif
```
I recommend `CFG_HWSUPP_MEM_PERM_WXN=n` and keep `CFG_CORE_RWDATA_NOEXEC=y`.
The later insures optee_os mapping splits executable and data, and the former does not enforce write-implies-noexec from SCTLR.

MrVan replied:
ok. I'll use CFG_HWSUPP_MEM_PERM_WXN ?= n in next version.

etienne-lms commented on core/arch/arm/plat-imx/pm/pm-imx7.c:
```diff
@@ -0,0 +1,264 @@
+	}
+
+	/* Note IRAM_S_BASE is not 1M aligned, so take care */
+	map.pa = ROUNDDOWN(IRAM_S_BASE, 0x100000);
```
minor: s/0x100000/CORE_MMU_PGDIR_SIZE/, here and below.

MrVan replied:
fix in next version.

etienne-lms commented on core/arch/arm/plat-imx/pm/pm-imx7.c:
```diff
@@ -0,0 +1,264 @@
+	uint32_t (*ddrc_phy_offset_array)[2];
+	uint32_t suspend_ocram_base = (uint32_t)phys_to_virt(
+			(paddr_t)TRUSTZONE_OCRAM_START +
+			SUSPEND_OCRAM_OFFSET, MEM_AREA_TEE_COHERENT);
```
Since this is coherent memory, we can expect phys=virt. Maybe use `phys_to_virt()` only under an `assert()`.

MrVan replied:
I could not get you. Could you please explain more?

jenswi-linaro commented on core/arch/arm/include/sm/pm.h:
```diff
@@ -0,0 +1,48 @@
+};
+
+/* suspend/resume core functions */
+void cpu_suspend_save(struct sm_pm_ctx *ptr, uint32_t save_sz, uint32_t sp);
```
The `save_sz` parameter isn't used and as far as I understand it will always be `sizeof(struct sm_pm_ctx)`

MrVan replied:
The save_sz maybe changed in future. I prefer we keep it now to fine tune cpu_suspend_save for the cache operation.

jenswi-linaro commented on core/arch/arm/include/sm/pm.h:
```diff
@@ -0,0 +1,48 @@
+	paddr_t cpu_resume_addr;
+	uint32_t suspend_regs[16];
+};
+
```
Please add a "sm_pm_" prefix to the functions.

MrVan replied:
ok. Will add them in next version patch

etienne-lms commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,235 @@
+END_FUNC cpu_resume_mmu
+
+/*
+ * Restore the registers stored when cpu_do_suspend
```
Add info `r0` shall points to the physical base address of the `suspend_regs` field of `struct sm_pm_ctx `.
Propose to add to description: `void cpu_do_resume(paddr suspend_regs) __noreturn;`.

MrVan commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,229 @@
+	push	{r4 - r12, lr}
+UNWIND(.save	{r4 - r12, lr})
+	mov	r5, sp
+	sub	sp, sp, #SM_PM_CTX_SIZE
```
I do not have much knowledge about unwind. Here it may lost. I would not like to change the design here.

jenswi-linaro replied:
We should replace `.save {r4 - r12, lr}` with `.cantunwind`

MrVan commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,235 @@
+
+FUNC cpu_resume_mmu, :
+UNWIND(.fnstart)
+	isb
```
write_sctlr r0 will turn on mmu. To keep safe, I prefer we keep isb here. cpu_resume_mmu here is to turn on mmu, and run in virtual address, before cpu_resume_mmu, all is runing in physical address. cpu_v7_suspend is common for armv7 here. But regarding a7/a9/a8/a15 and etc, cpu_do_suspend could handle. I think better to keep them.

MrVan commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,235 @@
+	isb
+	mov	r0, #0
+	/* Back to caller of cpu_suspend, with return value 0 */
+	pop	{r4 - r12, pc}
```
This is back to the next instruction just after saying `bl cpu_suspend`. The pop instruction in cpu_suspend is that cpu not suspended. Here the pop in cpu_resume path is cpu really suspened. They are in different code path. Let's not use jump to mix the two path.

MrVan commented on core/arch/arm/include/sm/pm.h:
```diff
@@ -0,0 +1,48 @@
+void cpu_do_resume(void);
+
+/*
+ * Exported to platform suspend, arg will be passed to fn as r0
```
will add it in next version.

etienne-lms commented on core/arch/arm/plat-imx/pm/psci-suspend-imx7.S:
```diff
@@ -0,0 +1,732 @@
+	str	r7, [r3, #DDRC_PWRCTL]
+
+	.endm
+
```
Could we add a comment to state that the routine `imx7_suspend` is copied to coherent memory and executed from there without any dependency on code/data in tee memory (DDR).

MrVan replied:
Have added comments in the beginning of this file:
```
+/*
 + * The code in this file is copied to coherent on-chip ram memory,
 + * without any dependency on code/data in tee memory(DDR).
 + */
```

MrVan commented:
@jenswi-linaro @etienne-lms Updated patchset, add comments for map_memarea_sections in headerfile, add cantunwind for most asm functions in this patchset, switch to add sm_pm_ prefix for almost all the functions used for suspend/resume, add more comments, replace 0x100000 with PGDIR size and MMU DEVICE SIZE, use CFG_HWSUPP_MEM_PERM_WXN = n. Thanks.

jenswi-linaro commented:
Quite a few of my comments has been ignored, please either address them or explain why you choose to ignore them.

MrVan commented:
@jenswi-linaro I just pushed a follow up patch, I just check the comments, I think only two comments are missed in previous patchset. Please check if this is ok. Thanks.

jenswi-linaro commented on core/arch/arm/sm/pm.c:
```diff
@@ -0,0 +1,72 @@
+#error "Max support 4 cores in one cluster now"
+#endif
+
+void sm_pm_cpu_suspend_save(struct sm_pm_ctx *ctx, uint32_t save_sz __unused,
```
Please remove the unneeded parameters. Right now it's only confusing as it's quite clear what size `*ctx` has, it doesn't cost anything to reintroduce it when needed.

MrVan replied:
ok. I'll remove the savesz.

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,243 @@
+FUNC sm_pm_cpu_resume_mmu, :
+UNWIND(.fnstart)
+UNWIND( .cantunwind)
+	isb
```
The `isb` is either needed or not needed, in this case it's not needed. It would be clearer if this code was at the end of `cpu_do_resume()`. Please fold `cpu_resume_mmu()` into the end of `cpu_do_resume()`.

What I'm asking for is better symmetry between the suspend and resume functions, If you'd rather add a `cpu_v7_resume()` than folding `cpu_v7_suspend()` into `cpu_do_suspend()` that's fine.

MrVan replied:
ok. I'll move cpu_v7_susend/resume into cpu_do_suspend/resume.

jenswi-linaro commented on core/arch/arm/sm/suspend.S:
```diff
@@ -0,0 +1,243 @@
+	isb
+	mov	r0, #0
+	/* Back to caller of sm_pm_cpu_suspend, with return value 0 */
+	pop	{r4 - r12, pc}
```
In both cases it will look like `cpu_suspend()` returned, I don't see a problem with executing the same code for different purposes. The only way for the caller to tell the difference is to check the return value.

By using the same pop instruction you avoid the problem with keeping two lines of code in sync. If you don't want to change the code the least you can do is adding comments noting the dependency.

MrVan replied:
I remove this and use the following as you suggested.
```
cpu_suspend:
xxxx
suspend_return:
   pop {r4 -  r12, pc}


cpu_resume:
xxxx
xxxx
b suspend_return.
````

jenswi-linaro commented on core/arch/arm/plat-imx/pm/psci-suspend-imx7.S:
```diff
@@ -0,0 +1,738 @@
+	str	r7, [r11, #0x38]
+wait_shutdown:
+	wfi
+	nop
```
Then note that the `nop` instructions are here for platform legacy reasons. Please add a comment at each place where you have one of more `nop` instructions.

The reason I'm asking for this is that this code could be copied into another platform and with the non-obvious `nop` instructions could make that code harder to maintain than necessary.

MrVan replied:
CPU would prefetch instructions. So add them to fill the pipeline. seems not needed, i'll try to remove them. I need to do a new round stress test in this weekend to make sure the whole flow works well. Changed a lot since v1 patchset:)

MrVan commented:
@jenswi-linaro  Updated. Remove the clk patch for ocotp reading, it is not neeed. Merged the v7_suspend/resume to cpu_do_suspend/resume. And Remove the nop in imx code. Remove savesz in cpu suspend save

MrVan commented on core/arch/arm/include/sm/psci.h:
```diff
@@ -31,6 +32,16 @@
+#define PSCI_POWER_STATE_ID_MASK	0xffff
+#define PSCI_POWER_STATE_ID_SHIFT	0
+#define PSCI_POWER_STATE_TYPE_SHIFT	16
+#define PSCI_POWER_STATE_TYPE_MASK	(1 << PSCI_POWER_STATE_TYPE_SHIFT)
```
ok, fix in next version

MrVan commented on core/arch/arm/plat-imx/pm/pm-imx7.c:
```diff
@@ -0,0 +1,264 @@
+	uint32_t i;
+	struct tee_mmap_region map;
+
+	/* iram tlb already initialized */
```
This is to build mmu table for suspend. When code run in iram, ddr is not accessible here.

jenswi-linaro replied:
Agree. It's the three letters in "tlb", how did you choose them?

MrVan replied:
sorry, @jenswi-linaro , I could not follow clearly. Do you mean "tlb" is not a good name here?

jenswi-linaro replied:
tlb for me is usually Translation Lookaside Buffer, but given a good explanation I'd accept it in this context too.

Or is it just a spell error that has been copied, was it supposed to be "tbl" from the beginning?

jbech-linaro replied:
`tbl` sounds more sane to me based on the discussion between the two of you. Also when it's in written format in the comment (not function names etc) why not even just write "table" instead of an abbreviation?

MrVan replied:
@jenswi-linaro @jbech-linaro "tbl" is better here, translation table. I did not find a good name, "mmu_table" too long, so picked "tlb" which caused misunderstanding. I'll replace tlb with tbl in this patch includeing functions and variables and add more comments.

MrVan commented on core/arch/arm/include/sm/psci.h:
```diff
@@ -31,6 +32,16 @@
+#define PSCI_POWER_STATE_TYPE_SHIFT	16
+#define PSCI_POWER_STATE_TYPE_MASK	(1 << PSCI_POWER_STATE_TYPE_SHIFT)
+#define PSCI_POWER_STATE_AFFL_SHIFT	24
+#define PSCI_POWER_STATE_AFFL_MASK	(0x3 << PSCI_POWER_STATE_AFFL_SHIFT)
```
Fix in next version

MrVan commented:
@jenswi-linaro Updated in this version. 
Renamed suspend.S to pm_a32.S; 
Use BIT32 and SHIFT_U32 in psci macro definitions; 
Change "tlb" to "tbl" in patch 10/10 core: arm: imx7d: add psci suspend support

Thanks for reviewing.

etienne-lms commented on core/arch/arm/sm/pm_a32.S:
```diff
@@ -0,0 +1,223 @@
+/*
+ * The following will be located in text section whose attribute is
+ * marked as readonly, but we only need to read here
+ * _suspend_sp stores the offset between thread_core_local to _suspend_sp
```
a word about `_core_pos` ?
or maybe just state that these allows `sm_pm_cpu_xxx` to call optee_os resources in a position independent manner.

MrVan replied:
ok. I'll add more comments here.

MrVan commented on core/arch/arm/plat-imx/pm/psci-suspend-imx7.S:
```diff
@@ -0,0 +1,720 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <kernel/unwind.h>
```
ok, fix in next version.

MrVan commented:
@etienne-lms I'll use a specific patch for moving definition of struct thread_core_local to thread.h.
About `For the generic part: suspend does no check other cpus have already suspended before deep suspending. I also wonder if some synchronisation between cpus are required.`. For suspend/resume case, before runs into suspend, core0 will use psci to kill secondary cores. For low power idle case, I use some synchoronisation code in platform specific code, not in common code.

glneo commented on core/arch/arm/sm/pm_a32.S:
```diff
@@ -0,0 +1,223 @@
+	write_sctlr r8
+	isb
+	mov	r0, #0
+	b	suspend_return
```
After reviewing these patches the only functions that are generic enough to be integrated into our(TI) suspend/resume implementation are 'sm_pm_cpu_do_suspend' and 'sm_pm_cpu_do_resume'.

One change we did have to make is this last line needed to be changed, we have to call into this function from another spot and so our stack is not in the same state as yours. If possible this function should be behave as a normal function (end with bx lr) so we can call it from our context.

If this patch could be isolated to just those functions I would like to take that and post our suspend/resume solution on top of that so we can compare/contrast and hopefully find more common ground.

MrVan replied:
@glneo Per @jenswi-linaro 's comments, I changed from pop {r4-r12, pc} to b suspend_return. Why would you split the functions into stanalone files? Could you share you code or example? Then we could see how to proceed. Before I continue, I would hear voice from @jenswi-linaro and @etienne-lms

jenswi-linaro replied:
We've been reviewing this for quite some time now and both me and @etienne-lms have no or only minor remarks left. Given that I'd rather get this merged soon.
After that we can continue and see how we can fit TI's suspend/resume into this. @glneo, I hope that's OK with you too.

etienne-lms replied:
agree

glneo replied:
I have no specific issue with this being pulled. My comment above was made assuming this whole set would not be merged for some time, with that it would have worked better for us to split out this patch and merge first so we could base our work on it, if the whole set is about to get taken then that would work also.

MrVan replied:
Thanks. I'll send out a new version to adrress @etienne-lms 's comments soon

MrVan commented:
@etienne-lms Updated. struct thread_core_local from thread_private.h to thread.h was done in a specific patch; r2 and r3 are not removed in psci-suspend-imx7.S at the beginning, they are not used now; add comments for _core_pos;  sorted the includes for psci-sus-end-imx7.S.
Tested on mx7dsabresd board.
Thanks for revewing.

MrVan commented on core/arch/arm/sm/pm_a32.S:
```diff
@@ -0,0 +1,224 @@
+	/* Invalidate icache to PoU */
+	write_iciallu
+	/* set reserved context */
+	write_contextidr ip
```
near the end of sm_pm_cpu_do_resume just after writing sctlr to enable mmu, there is an isb. Is that ok?

MrVan commented on core/arch/arm/sm/pm_a32.S:
```diff
@@ -0,0 +1,224 @@
+	isb
+	dsb
+	/* MMU will be enabled here */
+	write_sctlr r8
```
This was done psci-suspend-imx7.S, not in this pm_a32.S, is this ok?

etienne-lms replied:
From psci-suspend-imx7.S,
- if cores stay up (standby case): the sequence enables bp from `imx7_suspend`, then enables mmu here. I think bp should be enabled only after mmu is enabled (for cores requiring a manul enable of bp).
- In cpu shutdown case, cpu wakes to `resume`. In this case, it seems to me that both mmu and bp will be both enables from here  `write_sctlr r8`, without previous bp cache cleanup (`write_bpiall`).

MrVan replied:
If cores stay up, it will not run into cpu resume. In psci-suspend-imx7.S, there are two wfi. If cpu stay up, cpu will directly runs into the following instruction of wfi. 

If cpu shutdown, write_sctlr will restore branch predictor. In psci-suspend-imx7.S, line 645. If you think it is required to add a write_bpiall here before dsb/isb, I could add one here to handle the case that platform specific code missed to add write_bpiall in resume path.

glneo replied:
I think it makes the most sense to add the write_bpiall here as the branch predictor is enabled here. This will reduce the number of needed assumptions on the platform specific code.

MrVan commented on core/arch/arm/plat-imx/main.c:
```diff
@@ -62,9 +62,46 @@ static const struct thread_handlers handlers = {
+#ifdef ANATOP_BASE
 register_phys_mem(MEM_AREA_IO_SEC, ANATOP_BASE, CORE_MMU_DEVICE_SIZE);
+#endif
+#ifdef GICD_BASE
```
On i.MX8, there is GICD and GICR, no GICC used here. Since this piece code is shared to i.mx family. I seperated them.

etienne-lms replied:
ok

MrVan commented on core/arch/arm/plat-imx/pm/imx7_suspend.c:
```diff
@@ -0,0 +1,89 @@
+	if (!suspended_init) {
+		imx7_suspend_init();
+		suspended_init = 1;
+	}
```
I would not change it now in this patchset. Later with new patches, I could change it and do a new round stress test.

etienne-lms replied:
ok

MrVan commented on core/arch/arm/include/kernel/thread.h:
```diff
@@ -45,6 +45,33 @@
+
+#ifdef ARM64
+/*
+ * struct thread_core_local need to have alignment suitable for a stack
```
This is just moving code from thread_private.h to thread.h. Anyway I'll fix it.

MrVan commented on core/arch/arm/sm/pm_a32.S:
```diff
@@ -0,0 +1,224 @@
+UNWIND(.cantunwind)
+	push	{r4 - r11}
+	read_midr r4
+	ubfx	r5, r4, #4, #12
```
It is common to armv7. I think no need to introduce extra macros them.

etienne-lms replied:
Ok. It's minor. I think `#define MIDR_PN_MASK 0xfff`/`#define MIDR_PN_SHIFT 4` could be nice from **arm32.h**. However, this can be done later.

MrVan commented:
Update the commit messages according to @etienne-lms 's suggestion. Reordered the include files. need->needs in thread.h  Use ctx->suspend_regs to replace &ctx->suspend_regs[0] in pm.c.  using tab in UNWIND. Add a write_bpiall before enabling mmu in resume path. And a few minor fixes.

Thanks for reviewing.

MrVan commented:
It has been added in the new patchset.

2017-09-14 21:13 GMT+08:00 Andrew F. Davis <notifications@github.com>:

> *@glneo* commented on this pull request.
> ------------------------------
>
> In core/arch/arm/sm/pm_a32.S
> <https://github.com/OP-TEE/optee_os/pull/1729#discussion_r138888805>:
>
> > +	write_ttbcr r7
> +#endif
> +
> +	ldmia	r0, {r4 - r7}
> +	write_prrr r4
> +	write_nmrr r5
> +	write_vbar r6
> +	write_nsacr r7
> +
> +	write_actlr r9
> +	write_cpacr r10
> +	write_mvbar r11
> +	isb
> +	dsb
> +	/* MMU will be enabled here */
> +	write_sctlr r8
>
> I think it makes the most sense to add the write_bpiall here as the branch
> predictor is enabled here. This will reduce the number of needed
> assumptions on the platform specific code.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OP-TEE/optee_os/pull/1729#discussion_r138888805>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AECMaE8vP1hdZQRc0xlNuN-oLKXZh_4Wks5siSZigaJpZM4Oqhq->
> .
>

MrVan commented:
@jenswi-linaro @etienne-lms do you have more comments? Is this patchset ok to have your r-b/a-b tags?

jenswi-linaro commented on core/arch/arm/include/kernel/thread.h:
```diff
@@ -279,6 +280,8 @@ struct thread_handlers {
 void thread_init_primary(const struct thread_handlers *handlers);
 void thread_init_per_cpu(void);
 
+struct thread_core_local *thread_get_core_local(void);
```
This change should go into the "core: arm: kernel: make thread_core_local public" commit

jenswi-linaro commented on core/arch/arm/kernel/thread.c:
```diff
@@ -89,7 +89,7 @@
 struct thread_ctx threads[CFG_NUM_THREADS];
 
-static struct thread_core_local thread_core_local[CFG_TEE_CORE_NB_CORE];
+struct thread_core_local thread_core_local[CFG_TEE_CORE_NB_CORE];
```
This change should go into the "core: arm: kernel: make thread_core_local public" commit

jenswi-linaro commented on core/arch/arm/kernel/thread_private.h:
```diff
@@ -163,8 +163,6 @@ void thread_init_vbar(void);
 /* Handles a stdcall, r0-r7 holds the parameters */
 void thread_std_smc_entry(void);
 
-struct thread_core_local *thread_get_core_local(void);
```
This change should go into the "core: arm: kernel: make thread_core_local public" commit

MrVan replied:
Thanks @jenswi-linaro. I'll fix this, then add your tags.

MrVan commented:
"struct thread_core_local *thread_get_core_local(void);" changes moved to "core: arm: kernel: make thread_core_local public" commit.
Rebased to master
Added r-b/a-b tags

MrVan commented:
Missed one comment from @jenswi-linaro in last patchset:
```
-static struct thread_core_local thread_core_local[CFG_TEE_CORE_NB_CORE];
+struct thread_core_local thread_core_local[CFG_TEE_CORE_NB_CORE];
```
This change has been address in lastest patchset, with TAGS applied. Thanks.

MrVan commented on core/arch/arm/kernel/asm-defines.c:
```diff
@@ -55,6 +56,16 @@ DEFINES
+	DEFINE(THREAD_CORE_LOCAL_SIZE, sizeof(struct thread_core_local));
+
+	DEFINE(SM_PM_CTX_SP, offsetof(struct sm_pm_ctx, sp));
+	DEFINE(SM_PM_CTX_RESUME_ADDR,
```
ok, I'll remove them.

MrVan commented:
checkpatch warning fixed for commit3; uneeded SM_PM_CTX_xx macros removed.
Applied @etienne-lms 's r-b/a-b tags;

jforissier commented:
@MrVan thanks for this valuable contribution.

jbech-linaro commented:
> @MrVan thanks for this valuable contribution.

Indeed, thanks a lot for all the patience here!
